text/0002-serializable-types.md

```diff
@@ -0,0 +1,553 @@
+- Feature Name: `serializable-types`
+- Start Date: 2020-11-06
+- RFC PR: [RFC#2](https://github.com/Unity-Technologies/com.unity.multiplayer.rfcs/pull/2)
+- Issue: [MLAPI#454](https://github.com/Unity-Technologies/com.unity.multiplayer.mlapi/issues/454)
+
+# Summary
+[summary]: #summary
+
+This RFC proposes a standard way to support both built-in and custom serialization flows. Proposed design could be further extended with additional built-in serialization support, templated custom serializers and others in the future.
+
+# Motivation
+[motivation]: #motivation
+
+Type serialization is one of the most common areas in the network programming. We want to automate this process as much as we can, offer out-of-the-box features to make life easier.
+
+MLAPI used to offer serialization support that is much less performant (due to boxing, runtime type checking etc.) and (arguably) less convenient. When [Standard RPC API](https://github.com/Unity-Technologies/com.unity.multiplayer.rfcs/blob/master/text/0001-std-rpc-api.md) was introduced, it also did open a whole set of opportunities in this area. At the time of writing this RFC, this proposed design applies to the serialization of RPC parameters and is very likely to be used in other areas that use serialization such as network variables.
+
+We want to offer built-in serialization support for most commonly used types and also support custom serialization for user-defined types with ease.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## Serialization
+
+Multiplayer framework has built-in serialization support for C# and Unity primitive types out-of-the-box, also with ability to further extend network serialization for user-defined types implementing `INetworkSerializable` interface.
+
+### C# Primitives
+
+`bool`, `char`, `sbyte`, `byte`, `short`, `ushort`, `int`, `uint`, `long`, `ulong`, `float`, `double`, `string` types will be serialized by built-in serialization code.
```

Comment:
this one could be something we'd address later as it's called out in the "Future possibilities" section at the bottom:
> **Templated Custom Serializers**
> Currently, there is no way to swap `BitSerializer` with something else. It is still possible to manually serialize types down to `byte[]` and use `BitSerializer.Serialize(ref byte[] value)` API but it would be much better to have full control over serializer. This RFC considers this as a potential improvement for the future.

on your top comment, I'd like to touch on a few things.

first of all, BitStream and other managed types around serialization will be replaced with NativeArray and unmanaged memory operations which are also going to be helpful when we move more and more towards jobified internals. for now, we're just sticking with BitStream/BitReader/BitWriter trio and adding BitSerializer to help us just a little more. performance-wise, what we have today is not ideal and a bit far from what we're planning for the future.

for the other major areas: custom serialization, extending with new types, having more manual control and all other will potentially be introduced with a new API but not necessarily under existing BitSerializer/BitStream/BitReader/BitWriter stuff. `Serialize[Type]V2` API won't ever become a reality. what will be happening though is that we will introduce new set of classes backed by NativeArray/NativeSlice which are going to be job-friendly, more modular & customizable and potentially offer much much more value add compared to what we have today and what we will get with this RFC.

in terms of planning and rollout, we will most likely to have BitXXX stuff and new cool stuff coming up later side by side for a while, up until we fully migrate into cool stuff everywhere around the whole framework (including internals + public facing APIs) with 1:1 matching functionality at bare minimum and potentially more stuff on top.


again, long story short here: I 100% agree on all the points you raised above and I think we will eventually get there. only thing to call out here is that this RFC makes existing stuff a little bit better but won't be the final solution. we want these changes to improve things today and we will make further changes in this area again in the future.